pkg/kv/dynamodb/store.go

```diff
@@ -0,0 +1,320 @@
+package dynamodb
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/session"
+	"github.com/aws/aws-sdk-go/service/dynamodb"
+	"github.com/aws/aws-sdk-go/service/dynamodb/dynamodbattribute"
+	"github.com/treeverse/lakefs/pkg/kv"
+)
+
+type Driver struct{}
+
+type Store struct {
+	svc    *dynamodb.DynamoDB
+	params *Params
+}
+
+type EntriesIterator struct {
+	entry        *kv.Entry
+	err          error
+	store        *Store
+	queryResult  *dynamodb.QueryOutput
+	currEntryIdx int
+	partKey      string
+	startKey     string
+}
+
+type DynKVItem struct {
+	PartKey   string
+	ItemKey   string
+	ItemValue string
+}
+
+const (
+	DriverName       = "dynamodb"
+	DefaultTableName = "kvstore"
+
+	PartKey   = "PartKey"
+	ItemKey   = "ItemKey"
+	ItemValue = "ItemValue"
+
+	// TBD: Which values to use?
+	ReadCapacityUnits  = 1000
+	WriteCapacityUnits = 1000
+)
+
+//nolint:gochecknoinits
+func init() {
+	kv.Register(DriverName, &Driver{})
+}
+
+func (d *Driver) Open(ctx context.Context, name string) (kv.Store, error) {
+	sess := session.Must(session.NewSessionWithOptions(session.Options{
+		SharedConfigState: session.SharedConfigEnable,
+	}))
+
+	// Create DynamoDB client
+	svc := dynamodb.New(sess,
+		aws.NewConfig().
+			WithEndpoint(name))
+
+	// TODO: Get table name from env
+	params := &Params{TableName: DefaultTableName}
+	err := setupKeyValueDatabase(ctx, svc, params)
+	if err != nil {
+		return nil, fmt.Errorf("%w: %s", kv.ErrSetupFailed, err)
+	}
+
+	return &Store{
+		svc:    svc,
+		params: params,
+	}, nil
+}
+
+type Params struct {
+	TableName string
+}
+
+func (s *Store) getPartitionKey(_ /*itemKey*/ []byte) string {
+	// At some point the partition key will probably depend on the key.
+	// Meanwhile, it is simply the table name (i.e. everything is under the same partition)
+	return s.params.TableName
+}
+
+// setupKeyValueDatabase setup everything required to enable kv over postgres
+func setupKeyValueDatabase(ctx context.Context, svc *dynamodb.DynamoDB, params *Params) error {
+	// main kv table
+	_, err := svc.CreateTableWithContext(ctx, &dynamodb.CreateTableInput{
+		TableName: aws.String(params.TableName),
+		AttributeDefinitions: []*dynamodb.AttributeDefinition{
+			{
+				AttributeName: aws.String(PartKey),
+				AttributeType: aws.String("S"),
+			},
+			{
+				AttributeName: aws.String(ItemKey),
+				AttributeType: aws.String("S"),
+			},
+		},
+		KeySchema: []*dynamodb.KeySchemaElement{
+			{
+				AttributeName: aws.String(PartKey),
+				KeyType:       aws.String("HASH"),
+			},
+			{
+				AttributeName: aws.String(ItemKey),
+				KeyType:       aws.String("RANGE"),
+			},
+		},
+		ProvisionedThroughput: &dynamodb.ProvisionedThroughput{
+			ReadCapacityUnits:  aws.Int64(ReadCapacityUnits),
+			WriteCapacityUnits: aws.Int64(WriteCapacityUnits),
+		},
+	})
+	if err != nil {
+		if _, ok := err.(*dynamodb.ResourceInUseException); !ok {
```

Comment:
What's special about this error?

Reply:
This is the error DynamoDB returns if the table already exists. As such, it can be ignored and the creation con be considered successful